### PR TITLE
Use nvrtc for kernel compilation when available

### PR DESCRIPTION
**Pull request by abergeron**

### CODE REVIEW

**Comment by abergeron** [comment]
This gives a 2-3x speedup for kernel compilation.

File: src/gpuarray_buffer_cuda.c [hunk 1]
```diff
@@ -564,25 +521,126 @@ static CUresult get_cc(CUdevice dev, int *maj, int *min) {
...
+#ifdef WITH_NVRTC
+
+#include <nvrtc.h>
+
+static void *call_compiler(const char *src, size_t len, size_t *bin_len,
+                           char **log, size_t *log_len, int *ret) {
+  nvrtcProgram prog;
+  void *buf = NULL;
+  size_t buflen;
+  char compute_arg[12]; /* Must be 12. See detect_compute() */
```
Comment (nouiz) [inline src/gpuarray_buffer_cuda.c]: update the comment.

**Comment by nouiz** [comment]
I'm running it with cuda 7.0. I got error like this:

```pygpu.tests.test_blas.test_ger(4, 5, 'float32', 'f', 1, 1, False, True) ... ok
pygpu.tests.test_blas.test_ger(4, 5, 'float32', 'f', 1, 1, False, False) ... ok
pygpu.tests.test_elemwise.test_elemwise1_ops_array(<built-in function neg>, 'float32') ... ERROR
pygpu.tests.test_elemwise.test_elemwise1_ops_array(<built-in function neg>, 'float64') ... ERROR
pygpu.tests.test_elemwise.test_elemwise1_ops_array(<built-in function neg>, 'int8') ... ERROR
pygpu.tests.test_elemwise.test_elemwise1_ops_array(<built-in function neg>, 'int16') ... ^C
======================================================================
ERROR: pygpu.tests.test_elemwise.test_elemwise1_ops_array(<built-in function neg>, 'float32')
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Tmp/lisa/os_v4/anaconda/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/u/bastienf/.local/lib64/python2.7/site-packages/pygpu-0.2.1-py2.7-linux-x86_64.egg/pygpu/tests/support.py", line 39, in f
    func(*args, **kwargs)
  File "/u/bastienf/.local/lib64/python2.7/site-packages/pygpu-0.2.1-py2.7-linux-x86_64.egg/pygpu/tests/test_elemwise.py", line 36, in elemwise1_ops_array
    out_g = op(g)
  File "/u/bastienf/.local/lib64/python2.7/site-packages/pygpu-0.2.1-py2.7-linux-x86_64.egg/pygpu/_array.py", line 186, in __neg__
    return elemwise1(self, '-')
  File "/u/bastienf/.local/lib64/python2.7/site-packages/pygpu-0.2.1-py2.7-linux-x86_64.egg/pygpu/elemwise.py", line 522, in elemwise1
    k = ElemwiseKernel(a.context, args, oper)
  File "/u/bastienf/.local/lib64/python2.7/site-packages/pygpu-0.2.1-py2.7-linux-x86_64.egg/pygpu/elemwise.py", line 282, in __init__
    **self.flags)
  File "pygpu/gpuarray.pyx", line 1889, in pygpu.gpuarray.GpuKernel.__cinit__ (pygpu/gpuarray.c:21841)
  File "pygpu/gpuarray.pyx", line 451, in pygpu.gpuarray.kernel_init (pygpu/gpuarray.c:6387)
GpuArrayException: CUDA kernel build failure ::
0001	#define local_barrier() __syncthreads()
0002	#define WITHIN_KERNEL extern "C" __device__
0003	#define KERNEL extern "C" __global__
0004	#define GLOBAL_MEM /* empty */
0005	#define LOCAL_MEM __shared__
0006	#define LOCAL_MEM_ARG /* empty */
0007	#define REQD_WG_SIZE(X,Y,Z) __launch_bounds__(X*Y, Z)
0008	#ifdef NAN
0009	#undef NAN
0010	#endif
0011	#define NAN __int_as_float(0x7fffffff)
0012	#define LID_0 threadIdx.x
0013	#define LID_1 threadIdx.y
0014	#define LID_2 threadIdx.z
0015	#define LDIM_0 blockDim.x
0016	#define LDIM_1 blockDim.y
0017	#define LDIM_2 blockDim.z
0018	#define GID_0 blockIdx.x
0019	#define GID_1 blockIdx.y
0020	#define GID_2 blockIdx.z
0021	#define GDIM_0 gridDim.x
0022	#define GDIM_1 gridDim.y
0023	#define GDIM_2 gridDim.z
0024	#define ga_bool unsigned char
0025	#define ga_byte signed char
0026	#define ga_ubyte unsigned char
0027	#define ga_short short
0028	#define ga_ushort unsigned short
0029	#define ga_int int
0030	#define ga_uint unsigned int
0031	#define ga_long long long
0032	#define ga_ulong unsigned long long
0033	#define ga_float float
0034	#define ga_double double
0035	#define ga_half ga_ushort
0036	#define ga_size size_t
0037	#define ga_ssize ptrdiff_t
0038	
0039	
0040	
0041	KERNEL void elem_contig(const unsigned int n
0042	                    , GLOBAL_MEM ga_float * res
0043	                    , const unsigned int res_offset
0044	                    , GLOBAL_MEM ga_float * a
0045	                    , const unsigned int a_offset
0046	) {
0047	  const unsigned int idx = LDIM_0 * GID_0 + LID_0;
0048	  const unsigned int numThreads = LDIM_0 * GDIM_0;
0049	  unsigned int i;
0050	  GLOBAL_MEM char *tmp;
0051	
0052	  tmp = (GLOBAL_MEM char *)res; tmp += res_offset;
0053	  res = (GLOBAL_MEM ga_float *)tmp;
0054	  tmp = (GLOBAL_MEM char *)a; tmp += a_offset;
0055	  a = (GLOBAL_MEM ga_float *)tmp;
0056	
0057	  for (i = idx; i < n; i += numThreads) {
0058	    res[i] = -a[i];
0059	  }
0060	}

Compiler log:
default_program(61): warning: last line of file ends without a newline

default_program(61): error: unrecognized token

default_program(61): error: expected a declaration

default_program(61): error: unrecognized token

default_program(61): error: unrecognized token

default_program(61): error: unrecognized token

default_program(61): error: unrecognized token

default_program(61): error: unrecognized token

default_program(61): error: unrecognized token

default_program(61): error: unrecognized token

default_program(61): error: unrecognized token

10 errors detected in the compilation of "default_program".

```

**Comment by nouiz** [comment]
Do you know what could cause the errors I got?

**Comment by abergeron** [comment]
Apparently I forgot the final NUL in the source string, and it was the source of the problems.

**Comment by nouiz** [comment]
All tests passed locally with cuda 7. After my small comment about the name is done, we can merge.

**Comment by nouiz** [comment]
[bastienf@bart1 Build]$  make
Scanning dependencies of target gpuarray
[  3%] Building C object src/CMakeFiles/gpuarray.dir/gpuarray_types.c.o
[  6%] Building C object src/CMakeFiles/gpuarray.dir/gpuarray_error.c.o
[ 10%] Building C object src/CMakeFiles/gpuarray.dir/gpuarray_util.c.o
[ 13%] Building C object src/CMakeFiles/gpuarray.dir/gpuarray_buffer.c.o
[ 16%] Building C object src/CMakeFiles/gpuarray.dir/gpuarray_array.c.o
[ 20%] Building C object src/CMakeFiles/gpuarray.dir/gpuarray_array_blas.c.o
[ 23%] Building C object src/CMakeFiles/gpuarray.dir/gpuarray_kernel.c.o
[ 26%] Building C object src/CMakeFiles/gpuarray.dir/gpuarray_extension.c.o
[ 30%] Building C object src/CMakeFiles/gpuarray.dir/gpuarray_strl.c.o
[ 33%] Building C object src/CMakeFiles/gpuarray.dir/gpuarray_buffer_cuda.c.o
/u/bastienf/repos/libgpuarray/src/gpuarray_buffer_cuda.c:1623:38: error: ‘cuda_freekernel_pub’ undeclared here (not in a function)
make[2]: *** [src/CMakeFiles/gpuarray.dir/gpuarray_buffer_cuda.c.o] Error 1
make[1]: *** [src/CMakeFiles/gpuarray.dir/all] Error 2
make: *** [all] Error 2


Also need a rebase.

**Comment by abergeron** [comment]
Rebase done and problems fixed.

**Comment by nouiz** [comment]
WIth cuda 6.5 on a 580:
[bastienf@bart1 tests]$DEVICE=cuda0 nosetests test_blas.py -vx
*** Testing for GeForce GTX 580
pygpu.tests.test_blas.test_gemv((100, 128), 'float32', 'f', False, True, 1, True, False) ... ok
pygpu.tests.test_blas.test_gemv((100, 128), 'float32', 'f', False, True, 2, True, False) ... ERROR

======================================================================
ERROR: pygpu.tests.test_blas.test_gemv((100, 128), 'float32', 'f', False, True, 2, True, False)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Tmp/lisa/os_v4/anaconda/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/u/bastienf/.local/lib/python2.7/site-packages/pygpu-0.2.1-py2.7-linux-x86_64.egg/pygpu/tests/support.py", line 39, in f
    func(*args, **kwargs)
  File "/u/bastienf/.local/lib/python2.7/site-packages/pygpu-0.2.1-py2.7-linux-x86_64.egg/pygpu/tests/test_blas.py", line 40, in gemv
    sliced=sliced, ctx=context)
  File "/u/bastienf/.local/lib/python2.7/site-packages/pygpu-0.2.1-py2.7-linux-x86_64.egg/pygpu/tests/support.py", line 152, in gen_gpuarray
    assert numpy.allclose(a, numpy.asarray(b))
  File "/Tmp/lisa/os_v4/anaconda/lib/python2.7/site-packages/numpy/core/numeric.py", line 462, in asarray
    return array(a, dtype, copy=False, order=order)
  File "pygpu/gpuarray.pyx", line 1423, in pygpu.gpuarray.GpuArray.__array__ (pygpu/gpuarray.c:17000)
  File "pygpu/gpuarray.pyx", line 1241, in pygpu.gpuarray.pygpu_as_ndarray (pygpu/gpuarray.c:14900)
  File "pygpu/gpuarray.pyx", line 674, in pygpu.gpuarray.pygpu_copy (pygpu/gpuarray.c:8389)
  File "pygpu/gpuarray.pyx", line 411, in pygpu.gpuarray.array_copy (pygpu/gpuarray.c:5942)
GpuArrayException: No kernel image suitable for device

----------------------------------------------------------------------

**Comment by abergeron** [comment]
There was a problem with the PTX version in extcopy.  This should now work on any version of cuda > 6.0.  It may work with older versions, but this is not tested.

**Pull request merged**